File: 331_verify_preorder_serialization_of_a_binary_tree/solution.c
char *nextnode(char *node)
{
    for (node += 1; *node != '\0'; node++) {
        if (*node == ',')
            return &node[1];
    }
    return node;
}

char *sibling(char *node)
{
    char *left, *right;
    int right_child;
    if (*node == '\0')
        return NULL;
    if (*node == '#')
        return nextnode(node);
    left = nextnode(node);
    right = sibling(left);
    if (right == NULL)
        return NULL;
    return sibling(right);
}

bool isValidSerialization(char * preorder){
    if (*preorder == '\0')
        return true;
    preorder = sibling(preorder);
    if (preorder == NULL || *preorder != '\0')
        return false;
    return true;
}

File: 331_verify_preorder_serialization_of_a_binary_tree/solution.py
def is_valid(nrs, idx):
    # returns whether left and right node exists, and the idx of right child.
    if len(nrs) <= idx:
        return False, -1
    if nrs[idx] == '#':
        return True, idx + 1
    
    lcidx = idx + 1
    valid, ridx = is_valid(nrs, lcidx)
    if not valid:
        return False, -1
    valid, ridx = is_valid(nrs, ridx)
    if not valid:
        return False, -1
    return True, ridx

class Solution(object):
    def isValidSerialization(self, preorder):
        """
        :type preorder: str
        :rtype: bool
        """
        nrs = preorder.split(',')
        valid, ridx = is_valid(nrs, 0)
        if ridx == len(nrs):
            return True
        return False
